feat(firebase_functions): adding Firebase Functions package

TimWhiting opened the pull request:
Hi!

Glad to see an effort to reproduce the flutterfire interfaces in pure dart / for desktop.

Although I can't commit to contributing a lot, I did want to contribute something simple to show support for the community. Therefore I decided to attempt doing the cloud_functions plugin.

I've verified it working in pure dart locally with an emulator, but would appreciate any feedback you have. I assumed that you want the pure dart interface to be as similar as possible to the flutter interface.

The testing was mostly copied from the cloud_functions plugin. Tests in the desktop package are not complete, because I haven't worked a lot with mockito. Additionally the example is from the firebase_core_desktop plugin, and not updated to use firebase functions. The firebase configuration I took from the main flutterfire repo.

## TODOs

- [ ] Exception handling / formatting
- [ ] Auth integration
- [x] Using mocks in a few places in the tests (right now the fluttefire_functions_dart_test.dart relies on the emulator running)
- [ ] Converting some of the current tests to integration tests

Salakar commented:
> Although I can't commit to contributing a lot, I did want to contribute something simple to show support for the community. Therefore I decided to attempt doing the cloud_functions plugin.

This is a great contribution already @TimWhiting - thank you!

pr-Mais commented:
Hello @TimWhiting, thanks a ton for this valuable contribution!
Are you still working on it? if so, mind to convert it to a draft PR

TimWhiting commented:
Yeah, just converted to WIP PR. I'm fine with continuing to work on it / addressing feedback if you would like, or you can feel free to take it as is. I've updated the example, and tested the example on Linux. 

The outstanding TODOs currently are:
- [ ] Exception handling / formatting
- [ ] Auth integration
- [ ] Plugin endorsement (I saw you have a branch of flutterfire for that)
- [ ] Using mocks in a few places in the tests (right now the fluttefire_functions_dart_test.dart relies on the emulator running)
- [ ] Converting some of the current tests to integration tests

Let me know if you'd rather address these things in follow-on PRs as part of this project, or if you would like them addressed as part of the current PR.

pr-Mais commented:
Let's get the main package initially done in this PR, including Windows as well, then any subsequent issues or blocking cases can go in separate PRs
I will push the endorsement to the same branch on FF so you can test it as well

TimWhiting commented:
So by that you mean just getting the CI to pass for this PR? And then filing issues or adding things to the project board for the rest of the TODOs?

pr-Mais commented:
Currently the CI will not pass for windows, the problem should be fixed shortly, but I'm thinking Windows and Linux should both be tested before merging this PR

I currently use the project board with note cards, not issues since it's all still in progress, there's a board for Cloud Functions, you can use it to track all TODOs as well

cc @Salakar any other thoughts?

TimWhiting commented:
Okay, I can get Linux and Windows tested later this evening & improve the tests in general. I cannot add to the Cloud Functions board for those other items though.

pr-Mais commented:
Thank you
np then add a checklist to the description of this PR to track the TODOs

Salakar commented:
No major feedback from me - though I've not fully reviewed yet sorry. 

Only note from me is could we perhaps call the packages `firebase_functions*` if it's not too much effort - the ones on FF are historically not named correctly and eventually may be renamed to `firebase_functions-x` (same issue on `cloud_firestore` -> `firebase_firestore`) once ongoing refactors are finished.

TimWhiting commented:
Ready for review. I'll integrate auth, add integration tests, and provide better exceptions in a follow up PR, in addition to the endorsement of the plugin that needs to happen.

The Windows CI seems to have timed out. The tests are not integration tests and should pass on Windows the same as Linux. I've verified manually with the example app on both Linux and Macbook. Haven't got my windows machine set up for flutter desktop development yet though.

pr-Mais commented:
Currently, the plugins will not be endorsed, so we will go with [this approach](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#non-endorsed-federated-plugin) until they are stable.

pr-Mais commented:
@TimWhiting could you please rebase with `main`? probably it will get the Windows workflow to work.

pr-Mais commented:
All green 🎉

mikehardy commented on packages/firebase_functions/flutterfire_functions_desktop/LICENSE copy:
```diff
@@ -0,0 +1,29 @@
+BSD-3-Clause
```
LICENSE and "LICENSE copy" :thinking:

TimWhiting replied:
Yeah, I basically took firebase_core and modified it, and core had this. Do you want me to remove it from both places, or just here?

mikehardy replied:
I'm mostly a bystander on flutter related things but correctness seems to indicate there shouldn't be file copies anywhere. Not sure on removal process for firebase_core but at the least not committing a new copy here as an add seems correct

Salakar replied:
I've removed the 'copy' one from core on main - feel free to remove copy one here too for functions

mikehardy commented on packages/firebase_functions/flutterfire_functions_desktop/macos/flutterfire_functions_desktop.podspec:
```diff
@@ -0,0 +1,22 @@
+Pod::Spec.new do |s|
+  s.name             = 'flutterfire_functions_desktop'
+  s.version          = '0.0.1'
+  s.summary          = 'A new flutter plugin project.'
+  s.description      = <<-DESC
+A new flutter plugin project.
+                       DESC
+  s.homepage         = 'http://example.com'
+  s.license          = { :file => '../LICENSE' }
+  s.author           = { 'Your Company' => 'email@example.com' }
```
should the podspec contain boilerplate?

TimWhiting replied:
Not quite sure what to do here. The example app won't build unless there is a MacOS Pod, even though we declare the plugin as using a dartPluginClass and not using a pluginClass in the plugin's pubspec. I just generated this and the linux / windows plugin implementations from `flutter create -t plugin .`. They are basically no-ops and hopefully flutter in the future could just not require this if we have the dartPluginClass implementation. Happy to change this to whatever you want.

mikehardy replied:
oh interesting - I didn't realize this was really just box-ticking on a checklist essentially to make it work period. No change I think

mikehardy reviewed (changes requested):
Massive contribution, that's incredible! Awesome.
I had two comments that are best described as nitpicks, nevertheless are real I think

Salakar commented on packages/firebase_functions/flutterfire_functions_desktop/pubspec.yaml:
```diff
@@ -0,0 +1,47 @@
+  mockito: ^5.0.0
+  plugin_platform_interface: ^2.0.0
+
+dependency_overrides:
+  flutterfire_functions_dart:
+    path: ../flutterfire_functions_dart
```
This _shouldn't_ be required since Melos handles this - only inside ./example/pubspec.yaml is usually where we do overrides - was there an issue without this override?

TimWhiting replied:
Well, flutterfire_functions_dart has not been published yet (since it is in this PR), and I usually try to avoid having path dependencies in the regular dependencies (preferring to having path dependencies be in the overrides).

TimWhiting replied:
Related question, should the version number of theses firebase packages be in sync with each other or should this one start out at 0.0.1 or something else? I have it synced with core / auth right now.

mikehardy replied:
I may be in the minority here but I think semantic versioning is a thing and 0.n.n or 0.0.n starts things off in a way that specifically obscures important information for anyone that wants to consume it, as it squashes the type of change into more ambiguous semantic versioning buckets. Be bold @Salakar ! Start with 1.0.0, and let it fly free :laughing:

Salakar replied:
Let's start off at 0.1.0 - once all the packages are stable we'll graduate to 1.0.0 versioning;

----

@mikehardy Dart conventions (and Melos conventional commits supports bumping versions semantic-like pre 1.0.0 with this convention - until you explicitly tell it to graduate to 1.0.0) allow treating pre 1.0.0 semantically still ([source](https://dart.dev/tools/pub/versioning#semantic-versions));

> Although semantic versioning doesn’t promise any compatibility between versions prior to 1.0.0, the Dart community convention is to treat those versions semantically as well. The interpretation of each number is just shifted down one slot: going from 0.1.2 to 0.2.0 indicates a breaking change, going to 0.1.3 indicates a new feature, and going to 0.1.2+1 indicates a change that doesn’t affect the public API. For simplicity’s sake, avoid using + after the version reaches 1.0.0.

We will fly free, eventually 😅 I'd prefer to wait until we're sure the APIs are near enough stable and do a planned 1.0.0 graduation release in future since right now it's very much still in the experimental and figuring things out stage

mikehardy replied:
Ah, that's why there's so much `+` usage in Dart land. All the stuff is new (so lots of 0.x.y) but you still want the "standard 3" of break/feature/fix so it shifts down, but then the major of 0 does convey "hey, the API surface here may change completely. Makes sense :-)

pr-Mais commented on packages/firebase_functions/firebase_functions_desktop/pubspec.yaml:
```diff
@@ -16,7 +16,7 @@ dependencies:
     sdk: flutter
   flutterfire_core_dart: ^0.1.1-dev.0
-  flutterfire_functions_dart: ^1.0.0-dev.0
+  flutterfire_functions_dart: ^0.1.0-dev.0
```
I think we can go with naming it `firebase_functions_dart`, I originally started with this naming, e.g. `firebase_core_dart`, but this name for **core** and **auth** was taken so we had to come up with an alternative. I see it's not taken for functions.
@Salakar since `firebase_core_dart` and `firebase_core_auth` on pub are not maintained anymore, if we can take the names then let's publish the other packages with this naming style.

Salakar replied:
If we can get access to firebase_core_dart and firebase_core_auth then SGTM

pr-Mais replied:
Great, @TimWhiting can you please rename the dart package to `firebase_functions_dart`

pr-Mais commented:
Hello Tim, we've published both packages on pub, feel free to continue working on what's left, thanks again for the amazing work and contribution!

1. https://pub.dev/packages/firebase_functions_dart
2. https://pub.dev/packages/firebase_functions_desktop
